### ISSUE TITLE
No worky

### ISSUE THREAD

**Issue by ghost** [gfduszynski/cm-rgb#11]
Traceback (most recent call last):
  File "/usr/local/bin/cm-rgb-monitor", line 2, in <module>
    from cm_rgb.ctrl import LedChannel, LedMode, CMRGBController
ModuleNotFoundError: No module named 'cm_rgb'

Dependencies met, altered the primary script to point to Python3 as 2.7 is the default for the system XUbuntu 18.04

5.0.0-25-generic #26~18.04.1-Ubuntu

Did some poking, no dice. Lemme know if you need anymore info.

**Comment by gfduszynski** [comment]
What version of cm-rgb are you running ?
Did the previous version work for you ?

Did you install using pip or manually?

**Comment by ghost** [comment]
Holy CHIT that was fast...Uh there was a previous version? (just found out about this maybe a week ago) Based on my previous statements I'd hope it would be accurate to say I'm running the current/most recent build especially since I tried a reinstall just to see if the previous install has bombed on something.

pip install cm-rgb
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Requirement already satisfied: cm-rgb in /usr/local/lib/python2.7/dist-packages (0.0.9)
Requirement already satisfied: hidapi in /usr/local/lib/python2.7/dist-packages (from cm-rgb) (0.7.99.post21)
Requirement already satisfied: psutil in /usr/local/lib/python2.7/dist-packages (from cm-rgb) (5.6.7)
Requirement already satisfied: click in /usr/local/lib/python2.7/dist-packages (from cm-rgb) (7.0)
Requirement already satisfied: setuptools>=19.0 in /usr/local/lib/python2.7/dist-packages (from hidapi->cm-rgb) (44.0.0)

**Comment by gfduszynski** [comment]
Current version is 0.1.0, but that is not the cause.
I've tested this on python 3.0, is it possible you could upgrade to 3.0?

**Comment by gfduszynski** [comment]
Maybe you executed pip install with 2.7
and now you are executing with 3.0?

**Comment by ghost** [comment]
Py isn't my lang of choice thus forgive my ignorance but I can't find any way to install pip for py3 or to run the system install version with the installed py3. Pretty sure 18.04 EOL is a few years after the py2 EOL so not sure if there will be a migration planned or what but I suppose I'm SOL (see what I did there) until I move to 20.04 unless you can educate me.

Some searching said py(version) -m pip to select the py to use but I just get no module warnings (/usr/bin/python3.6: No module named pip). I have py3 installed but the only version of pip that will execute are py2 based and I can't find pip built on or for py3x.

You might want to denote py3 as a dependency so the installer will bail without. Less time supporting people like me heh. Plenty of 16.04 machines out there much less 18.04. Perhaps there will be an update to migrate LTS installs to py3 soon but I don't personally know. That said we don't know if py3 will change anything. I'm not sure what you have tested on. Could be some dependency only installed by a few full distros and missing on things like arch, lubuntu etc. I've seen a lot of that lately.

Had a bit more time to screw around. Reinstalled pip for what should be py3.6 along with a slight dependency hell. Failed, same issue. Reedited the cm-rgb script to call specifically 3.6...no change.

Anyhow don't worry about it. If others start to report then go to town. I'll check back in on 20.04. I appreciate the work though for sure. I can't wait for this RGB vomit fest trend to be some meme joke about remember back in the day when your computer would give your epileptic cousin seizures?...WORD...

**Comment by gfduszynski** [comment]
Well to be fair this cli is the first thing i wrote in python so I'm not an expert either 😄 

Correct me if I'm wrong, you currently have python 3.6 installed with pip ?

https://linuxize.com/post/how-to-install-pip-on-ubuntu-18.04/
According to the above there should be package ready for pip for py3.
Did you install pip through package or some other means ?

**Comment by gfduszynski** [comment]
Try to follow this:
https://stackoverflow.com/questions/40834656/installed-module-using-pip-not-found

There was another potential issue with installing with sudo:
https://askubuntu.com/questions/1069494/python-module-not-found-when-launched-in-cron

**Comment by ghost** [comment]
Shhh you are supposed to instill faith in your work, at least fake it until you make it :P I hear you. I've dabbled in python over the years but the things I tried were broke as helll (glaring at you PyGame). I'd like to say I'd dig in more but I have some huge projects on my plate. I just came across an article talking about this and was like YES RGB puke BEGONE! heh. Though honestly I do like the idea of the LEDS being able to show CPU load but I'd never really see anything since the machine is behind my desk on my server rack behind a rather literal wall of 13 monitors. I've been wanting to upgrade but the thought of paying extra for RGB crap simply because the "good" stuff like RAM is all RGB now makes me cringe but this is highly off topic.

Back on topic do you mind me asking what you are running? Plain Jane Debian? Flavor or Ubuntu? Arch? Something more obscure? (lord thinking now I am old...cue Slackware) Fedora? Suse? Just curious about what the python base is. Given Py2 is about to be EOL'd with an axe I'd imagine most have switched to 3. Kinda surprised 18.xx shipped with 2 but I know the migration process is hell but time is running out.

That said you asked a bit ago if I installed with pip or manual and the answer is both. Both yield the same results/behaviors.

**Comment by ghost** [comment]
Tried with that article, different but same...
Traceback (most recent call last):
  File "/usr/local/bin/cm-rgb-monitor", line 58, in <module>
    monitor()
  File "/home/$USER/.local/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/$USER/.local/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/$USER/.local/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/$USER/.local/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/bin/cm-rgb-monitor", line 14, in monitor
    c = CMRGBController()
  File "/home/$USER/.local/lib/python3.6/site-packages/cm_rgb/ctrl.py", line 50, in __init__
    self.__init_hid_device()
  File "/home/$USER/.local/lib/python3.6/site-packages/cm_rgb/ctrl.py", line 57, in __init_hid_device
    raise Exception("No devices found")
Exception: No devices found


perhaps the usb issue...
cm-rgb-cli add-udev-rule
Usage: cm-rgb-cli [OPTIONS] COMMAND1 [ARGS]... [COMMAND2 [ARGS]...]...
Try "cm-rgb-cli --help" for help.

Error: No such command "add-udev-rule".

I checked the post on the sudo pip issue but that's kinda inverted and I have tried both ways.

Forgive me if I'm a tad scattered, debugging my own stuff across a few machines atm.

**Comment by gfduszynski** [comment]
I'm on arch and you are almost there :)

add-udev-rule is present in v0.1.0

Just for test try running ``sudo cm-rgb-monitor``
You probably did not have access to usb device.

**Comment by ghost** [comment]
sudo cm-rgb-cli add-udev-rule
Usage: cm-rgb-cli [OPTIONS] COMMAND1 [ARGS]... [COMMAND2 [ARGS]...]...
Try "cm-rgb-cli --help" for help.

Error: No such command "add-udev-rule".
$USER@pillar:~/Desktop$ sudo cm-rgb-monitor --cpu-color=#FFA500 --bg-color=#00FFFF
Traceback (most recent call last):
  File "/usr/local/bin/cm-rgb-monitor", line 58, in <module>
    monitor()
  File "/home/$USER/.local/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/$USER/.local/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/$USER/.local/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/$USER/.local/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/bin/cm-rgb-monitor", line 14, in monitor
    c = CMRGBController()
  File "/home/$USER/.local/lib/python3.6/site-packages/cm_rgb/ctrl.py", line 50, in __init__
    self.__init_hid_device()
  File "/home/$USER/.local/lib/python3.6/site-packages/cm_rgb/ctrl.py", line 57, in __init_hid_device
    raise Exception("No devices found")
Exception: No devices found

**Comment by gfduszynski** [comment]
Do you have usb cable connected to your cooler ?
Check with ``lsusb`` if any entry looks like ``Cooler Master``

**Comment by ghost** [comment]
nope doesn't show but my header is right & all pwm is fine.

**Comment by gfduszynski** [comment]
This cooler uses separate cable to manage fan pwm (non detachable) and either RGB header cable (not supported by this cli) or USB cable.
![obraz](https://user-images.githubusercontent.com/1913102/72271148-a863ee80-3626-11ea-887d-4fef2e154bdb.png)

Some random image from the internet's with both RGB header and USB cables connected (not recommended)

**Comment by ghost** [comment]
Wording is confusing but I have 3 cables. Once my cat is done using my kybrd as bed I will do some reading.

**Comment by gfduszynski** [comment]
It's important to not have the cable for RGB header connected for this to work only the USB one.
Let me know if you succeed :)

**Comment by ghost** [comment]
Hate on scr keybrd but....forgot about these cables, it's been a while & I omitted because ryzen master & all that didn't work under wine so why waste the header. I'll power down in a bit & try...once I've dried & soaked up the drool from my keyboard...meowlyfe!

**Comment by ghost** [comment]
Works but will not run without root. That said what a bag of dicks that cable is on my mobo. When I plugged that BS in all the sudden the whole back of my mobo lit up, like wtf where did all this come from? COMPUTER != RAVE !!! Hoping I can set this and it will be saved in BIOS so I can undo that cable and forget it again.

udev rule fails Failed to create /etc/udev/rules.d/60-cm-rgb.rules. You also need to `type cm-rgb-cli` to get your path to the bin because the path isn't avail to root but even with the correct path it will fail still.

**Comment by gfduszynski** [comment]
xD

**Comment by gfduszynski** [comment]
Try `` pip install --upgrade cm-rgb ``
and then if you successfully get 0.1.0 try adding udev rule :)

**Comment by gfduszynski** [comment]
I know it's not pretty but you can try to add this rule manually
```
SUBSYSTEM=="usb", ATTR{idVendor}=="2516", ATTR{idProduct}=="0051", TAG+="uaccess", TAG+="udev-acl"
```

**Comment by ghost** [comment]
pip3 :P I don't want to be overly pedantic but someone else whose running an older distro may need to know pip might run py2.7 still.

Did the update, still fails. Also trying to use the set command fails a la

cm-rgb-cli set logo --mode=breathe --color=#00ff00 --speed=3 --brightness=5 fan --mode=breathe --color=#0000ff --speed=4 --brightness=1 ring --mode=swirl --color=#ff0000 --speed=1 --brightness=1 save

**Comment by ghost** [comment]
Lol nothing is pretty with udev, that said I've never had anything udev related work for me on any hardware. udev is like a spring loaded box of douche bag leprechauns.

**Comment by gfduszynski** [comment]
I'll have to return to this tomorrow, we are close though 👍

**Comment by ghost** [comment]
I manually added the rule so we'll see, I'll leave that cable for now until you can get back to seeing why the set option makes the cm-rgb-cli fail. I'll play for a few in the mean time before I get back to my own flustercluck of code. Thanks for your time man.

**Comment by ghost** [comment]
While I know there were other things at play I'd say it would also be good measure for you to build in some lsusb grep awk fun to ensure someone like me has the cable hooked up and showing to the system. I'd wager I'm not the only one that said F that stuff it won't work anyhow when we built never guessing anyone would do this. Again not an RGB fan so not up to speed with all this. I'd rather they leave this stuff out, make it cheaper, or cool better...you know the thing that matters?

**Comment by gfduszynski** [comment]
Does the ``set`` command still fail with ``Exception: No devices found`` ?

lsusb was just to be certain, scripts checks for the presence of the usb device.
Message could be more verbose I suppose.

**Comment by ghost** [comment]
cm-rgb-cli set logo --mode=breathe --color=#00ff00 --speed=3 --brightness=5 fan --mode=breathe --color=#0000ff --speed=4 --brightness=1 ring --mode=swirl --color=#ff0000 --speed=1 --brightness=1 save

If I remove set it seems to work but the syntax shown in your examples use set which causes the failure.

Usage: cm-rgb-cli [OPTIONS] COMMAND1 [ARGS]... [COMMAND2 [ARGS]...]...
Try "cm-rgb-cli --help" for help.

Error: No such command "set".

So this works(ish)
cm-rgb-cli logo --mode=breathe --color=#00ff00 --speed=3 --brightness=5 fan --mode=breathe --color=#FFFF00 --speed=4 --brightness=1 ring --mode=swirl --color=#FFB60000 --speed=1 --brightness=1 save

**Comment by gfduszynski** [comment]
Ok, that means you have 0.0.9 most likely and you don't have to bother updating.
0.1.0 was a refactor with add-udev-rule command added.

Command set was added to avoid problems with argument parsing where you could mix save and restore command in one call etc.

**Comment by ghost** [comment]
OK? *shifty eyes*

Lol so much for checking the version...

cm-rgb-cli --help
Usage: cm-rgb-cli [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  add-udev-rule
  restore
  set
  version
shed@pillar:~$ cm-rgb-cli version
Traceback (most recent call last):
  File "$HOME/.local/bin/cm-rgb-cli", line 151, in <module>
    main_group()
  File "$HOME/.local/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "$HOME/.local/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "$HOME/.local/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "$HOME/.local/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "$HOME/.local/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "$HOME/.local/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "$HOME/.local/bin/cm-rgb-cli", line 131, in version
    print(CMRGBController().getVersion())
  File "$HOME/.local/lib/python3.6/site-packages/cm_rgb/ctrl.py", line 50, in __init__
    self.__init_hid_device()
  File "$HOME/.local/lib/python3.6/site-packages/cm_rgb/ctrl.py", line 60, in __init_hid_device
    self.device.open_path(device_list[0]["path"])
  File "hid.pyx", line 72, in hid.device.open_path
OSError: open failed

**Comment by gfduszynski** [comment]
:D that's unfortunate, I'll add that to my list of pending fixes.

**Comment by ghost** [comment]
Heh, I know the feeling. Thanks for your help and time. My machine is less Rave Tastic now thanks to you. Nothing against raves, I rather miss them but that's another story.
